Implemented statistics tracking

bartdp1 opened the pull request:
## What does this PR do?

self.evaluated_individuals is now a dict with key being the string representation of the individual and the value being a dict containing statistics about the pipeline at the time it was evaluated.

The DEAP individuals have a new attribute `statistics`, which is a dict containing current information about the individual. Initialization is done with a new function `initialize_stats_dict`, which is called in the `eaMuPlusLambda` function. 

Statistics are updates as individuals advance to the next generation, are mutated or crossed. More specifically

- generation - this is updated in the `eaMuPlusLambda` function as the iteration for a new generation is started.
- crossover_count - this is updated in the `self._mate_operator` function.
- mutation_count - this is updated in the `self._random_mutation_operator` function.
- predecessor - this is updated both the `self._mate_operator` and  `self._random_mutation_operator` 

a new function in the base class is introduced `self._final_stats` which consolidates the mentioned statistics with the operator count and cross validation score, and is called at any occasion where `self.evaluated_individuals` is updated. 

The `self_evaluate_individuals` function had some minor changes in order to return the right statistics.

`self._preprocess_individuals` returns an extra dict, where key is the string representation of the individuals, and the value is the dict containing the statistics. `_update_evaluated_individuals_` has this dict added to it's arguments. 

furthermore, the tests in tpot_tests.py are adjusted to account for the changes. stats_test.py is added to test the specific functionality regarding the stats.

## Where should the reviewer start?
Follow the changes as mentioned above.


## How should this PR be tested?
Tests for the added functionality can be found in `stats_test.py.` Suggestions for tests are welcome.



## Any background context you want to provide?



## What are the relevant issues?

#585

## Screenshots (if appropriate)



## Questions:

- Do the docs need to be updated?
Description of the evaluated_individuals_ should be adjusted.
- Does this PR add new (Python) dependencies?
No.

Commit:
Implemented statistics tracking
self.evaluated_individuals is now a dict with key being the string representation of the individual and the value being a dict containing statistics about the pipeline at the time it was evaluated.

rhiever commented:
I see. This is a pretty smart way to track lineage and related stats about each individual. I suppose since we don't really allow the same pipeline to be generated via crossover/mutation once it's been previously evaluated (barring the situation where a crossover/mutation can't generate an entirely new individual within 10 attempts), there's really no need to maintain pointers from every offspring to its parents.

Well done and thank you for the PR, @bartdp1. I'll allow @weixuanfu some time to review this PR as well. I've added my comments.

rhiever commented on tpot/base.py:
```diff
@@ -1059,9 +1066,14 @@ def _evaluate_individuals(self, individuals, features, target, sample_weight=Non
-        self._update_evaluated_individuals_(result_score_list, eval_individuals_str, operator_counts)
+        self._update_evaluated_individuals_(result_score_list, eval_individuals_str, operator_counts, stats_dicts)
+
+        def get_optimization_values(individual):
```
`get_optimization_values` should be an internal function to TPOT, no? Otherwise it's constantly being re-defined every time `_evaluate_individuals` is being called.

rhiever replied:
Actually, is this function needed at all? It's only used once, below.

bartdp1 replied:
The function is indeed only used once below, but i find the list compresension in which the function is called better readable in this way. Will change it

rhiever commented on tpot/gp_deap.py:
```diff
@@ -218,6 +233,10 @@ def eaMuPlusLambda(population, toolbox, mu, lambda_, cxpb, mutpb, ngen, pbar,
 
+        # Update generation statistic for all individuals
+        for ind in population:
+            ind.statistics['generation'] = gen
```
This update procedure might be problematic for individuals that stick around from the previous generation. In our current GP algorithm, we don't have complete turnover between each iteration of the GP.

Need to add a check that each individuals' `generation` statistic hasn't already been set to a non-default value.

weixuanfu replied:
What is this `ind.statistics['generation']` use for? I think some description about it is needed. If it is about which generation(s) has this individual, it should be a list instead.

rhiever replied:
I believe the `generation` statistic is meant to track the generation that the individual first appeared.

Good point though, @weixuanfu. Since we're moving forward with this change, this PR should add some documentation (in the docstrings and the API documentation) about these new statistics.

bartdp1 replied:
Since the information in statistics only gets written to `evaluted_individuals_` upon evaluation, and an individual is evaluated in the generation that it first occurs, the statistics reported in `evaluated_individuals_` is always the information of the individual at the time of evaluation. In this light i did not build in any checks here, however, for debugging purposes it might be a good idea to indeed update `generation` only when it concerns a new individual, and not update for individuals that traverse from the previous generation. 

I can implement this if you like, let me know.

rhiever replied:
Because the behavior you described is a function of other parts of TPOT that are not directly related to the statistics tracking, and that behavior could change at any point and break this assumption, let's add an explicit check here to make sure that persistent individuals don't have their `generation` statistic updated. Please also add a comment above that check describing why we're doing that, so future developers aren't confused by the logic.

bartdp1 replied:
@rhiever Got it. How about incrementing the generation statistic along with the other statistics, at the moment the new individual is generated by mutation/cross-over? @rhiever. Am i missing any cases this way?

rhiever replied:
Yes, we could do that as long as the `generation` is available to that function.

rhiever replied:
Yes, we could do that as long as the `generation` is available to that function.

bartdp1 replied:
Ah i see, it's not Always generation = generation + 1. The way i've implemented the check:
Mutation and crossover 'invalidate' the generation stat. After varOr is run, the generation stat is updaten when invalid.

rhiever commented on tpot/gp_deap.py:
```diff
@@ -199,6 +210,10 @@ def eaMuPlusLambda(population, toolbox, mu, lambda_, cxpb, mutpb, ngen, pbar,
     logbook = tools.Logbook()
     logbook.header = ['gen', 'nevals'] + (stats.fields if stats else [])
 
+    # Initialize statistics dict
```
More descriptive comment, please.

rhiever commented on tpot/gp_deap.py:
```diff
@@ -149,6 +149,17 @@ def varOr(population, toolbox, lambda_, cxpb, mutpb):
+    '''
+    Initializes the stats dict for individual
+    :param individual:
+    :return:
```
Please make sure the function docstrings match the docstring format that we use in the rest of TPOT.

bartdp1 replied:
Will do. Is this any established format, then i can set my IDE settings which will save me some time.

bartdp1 replied:
Moreover, there is some discrepancy in the format used. E.g. comparing `mutate_random_individual` and `varOr` in `gp_deap`.

rhiever commented on tpot/base.py:
```diff
@@ -1230,6 +1257,10 @@ def _random_mutation_operator(self, individual, allow_shrink=True):
             ind = self._toolbox.clone(individual)
             offspring, = mutator(ind)
             if str(offspring) not in self.evaluated_individuals_:
+                # Update statistics
```
More descriptive comment, please.

rhiever replied:
Ping on this small change request, @bartdp1.

bartdp1 replied:
Missed it, Will do tomorrow

bartdp1 replied:
It's already there, in one of the last commits :)

rhiever commented on tpot/base.py:
```diff
@@ -1184,8 +1203,16 @@ def _mate_operator(self, ind1, ind2):
+                # update statistics
+                offspring.statistics['predecessor'] = (str(ind1), str(ind2))
+                offspring.statistics['mutation_count'] = ind1.statistics['mutation_count'] + ind2.statistics[
+                    'mutation_count']
```
No need for line break here; we use a lax version of the PEP8 guidelines.

bartdp1 replied:
Ok. Think this will just be flipping a switch in the IDE.

rhiever commented on tpot/base.py:
```diff
@@ -1184,8 +1203,16 @@ def _mate_operator(self, ind1, ind2):
+                offspring.statistics['mutation_count'] = ind1.statistics['mutation_count'] + ind2.statistics[
+                    'mutation_count']
+                offspring.statistics['crossover_count'] = ind1.statistics['crossover_count'] + ind2.statistics[
+                    'crossover_count'] + 1
```
No need for line break here; we use a lax version of the PEP8 guidelines.

rhiever commented on tpot/base.py:
```diff
@@ -1001,6 +1001,13 @@ def _stop_by_max_time_mins(self):
             if total_mins_elapsed >= self.max_time_mins:
                 raise KeyboardInterrupt('{} minutes have elapsed. TPOT will close down.'.format(total_mins_elapsed))
 
+    def _final_stats(self, operator_count, cv_score, individual_stats):
```
Maybe a more descriptive function name: `_combine_individual_stats`?

rhiever commented on tpot/base.py:
```diff
@@ -1001,6 +1001,13 @@ def _stop_by_max_time_mins(self):
                 raise KeyboardInterrupt('{} minutes have elapsed. TPOT will close down.'.format(total_mins_elapsed))
 
+    def _final_stats(self, operator_count, cv_score, individual_stats):
+        """Combine the stats with operator count and cv score and preprare to be written to _evaluated_individuals"""
```
Please annotate the parameters and output(s) of this function in the docstrings.

rhiever commented on tests/stats_test.py:
```diff
@@ -0,0 +1,104 @@
+from tpot import TPOTClassifier
```
Thank you for the unit tests. 👍

rhiever commented on tests/stats_test.py:
```diff
@@ -0,0 +1,104 @@
+        offspring1, offspring2 = tpot_obj._mate_operator(ind1, ind2)
+
+        assert offspring1.statistics['crossover_count'] == ind1.statistics['crossover_count'] + ind2.statistics[
+            'crossover_count'] + 1
```
No need for this newline.

rhiever commented on tests/stats_test.py:
```diff
@@ -0,0 +1,104 @@
+        assert offspring1.statistics['crossover_count'] == ind1.statistics['crossover_count'] + ind2.statistics[
+            'crossover_count'] + 1
+        assert offspring1.statistics['mutation_count'] == ind1.statistics['mutation_count'] + ind2.statistics[
+            'mutation_count']
```
No need for this newline.

weixuanfu commented on tpot/gp_deap.py:
```diff
@@ -149,6 +149,17 @@ def varOr(population, toolbox, lambda_, cxpb, mutpb):
+    '''
+    Initializes the stats dict for individual
+    :param individual:
+    :return:
```
Please add more detailed docstring here too.

weixuanfu commented on tpot/base.py:
```diff
@@ -1103,7 +1116,7 @@ def _preprocess_individuals(self, individuals):
             sklearn_pipeline_str = generate_pipeline_code(expr_to_tree(individual, self._pset), self.operators)
             if sklearn_pipeline_str.count('PolynomialFeatures') > 1:
-                self.evaluated_individuals_[individual_str] = (5000., -float('inf'))
+                self.evaluated_individuals_[individual_str] = self._final_stats(5000., -float('inf'), individual.statistics)
```
Please reformat this line as the line 1144.

weixuanfu commented on tpot/base.py:
```diff
@@ -1125,17 +1138,21 @@ def _preprocess_individuals(self, individuals):
 
 
-    def _update_evaluated_individuals_(self, result_score_list, eval_individuals_str, operator_counts):
+    def _update_evaluated_individuals_(self, result_score_list, eval_individuals_str, operator_counts, stats_dicts):
```
Please add more detailed docstring here.

bartdp1 committed:
Reformatting
Documentation
removes get_optimization_values function, return results directly

bartdp1 committed:
Refactor method and docs
_final_stats refactored to _combine_individual_stats
added missing documentation

bartdp1 committed:
Refactor method and docs
_final_stats refactored to _combine_individual_stats
added missing documentation

bartdp1 committed:
Comments

rhiever commented on tpot/base.py:
```diff
@@ -630,11 +625,11 @@ def _update_top_pipeline(self):
+                                   'not formatted properly, or because data for '
+                                   'a regression problem was provided to the '
+                                   'TPOTClassifier object. Please make sure you '
+                                   'passed the data to TPOT correctly.')
```
Thank you for the various formatting changes. However, as a note for future PRs, beware of making numerous small changes like these ones when they are unrelated to the core PR. Numerous small, unrelated changes can cause merge conflicts if another PR happens to be changing those lines as well. I think we can make it work this time.

bartdp1 replied:
Hmm, sorry for this. Think i might have accidently hit ‘auto format’ in my ide. Hopefully wont be harmfull in any way. Let me know if i should roll back Some things.

rhiever replied:
No worries. I'm pretty sure we can sort out any merge conflicts this time because there's only one other PR being actively developed (just merged into the dev branch).

bartdp1 committed:
Generation stat only gets updated when the individual is 'new'.

bartdp1 committed:
Merge branch 'development' into my-contributions

bartdp1 commented:
@rhiever @weixuanfu. Seems like the PR times out on some unit test. Found issue #588, which adresses the same problem. How can i best proceed with integrating the fix? Copying it and comitting it to the PR? Thanks in advance

weixuanfu commented:
@bartdp1 you may try the git commands below for updating these expensive unit tests to fix this issue.

```
git cherry-pick 68e42f2136ec43200964d355c5c68288af85e7f5
git cherry-pick 9b428d73e5724ecb61bebc40f69b8628ecc2474a
```

bartdp1 committed:
quick PolynomialFeatures test

weixuanfu commented:
You may need [this commit](https://github.com/rhiever/tpot/pull/594/commits/9b428d73e5724ecb61bebc40f69b8628ecc2474a)

bartdp1 committed:
Expensive unit test speed tests

bartdp1 commented:
@rhiever @weixuanfu. I think i got all your feedback integrated now, please let me know if there are some more modifications i can make :).

rhiever commented:
Merged. Thanks again for your great PR, @bartdp1!
